Transform logic to extract signal debug names

AleksanderBodurri opened the pull request:
Implements a compiler transform that attempts to statically analyze variable names and apply them to usages of signal functions like signal, computed, effect, etc as debug names.

Related: #57074 

Depends on: #57073

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+    return false;
+  }
+
+  return importDeclaration.moduleSpecifier.text === '@angular/core';
```
Suggestion: Do `specifier === '@angular/core' || specifier.startsWith('@angular/core/')` so we can cover subpath exports as well.

I'm also wondering if we should do `specifier.startsWith('@angular/')` to support any package? If so, we'd run the risk of transforming any API named `signal`, `computed`, etc. from any Angular package, which might have different semantics (ex. `signal` from `@angular/common/http` might refer to something related to `AbortSignal`). So I think I'm leaning away from this but figured I'd mention it in case you feel differently?

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+    'contentChild',
+    'contentChildren',
+    'effect',
+  ]);
```
Optimization: Move to top-level scope outside `isSignalFunction` so we don't re-create this set for every call.

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+function isSignalFunction(expression: ts.Identifier) {
+  const text = expression.text;
+
+  const signalFunction = new Set([
```
Consider: `Object.freeze` this value and/or type it as `ReadonlySet` to protect against mutations at runtime.

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+  return importDeclaration.moduleSpecifier.text === '@angular/core';
+}
+
+function isSignalFunction(expression: ts.Identifier) {
```
Consider: Use explicit return types throughout this file.

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+    'viewChildren',
+    'contentChild',
+    'contentChildren',
+    'effect',
```
Consider: Could we make this a map of signal function -> a metadata object containing the config index for that function? Would that make `getConfigArgPosition` a bit less magical / contextual?

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+import ts from 'typescript';
+
+function insertDebugNameIntoExistingConfigObject(
```
Suggestion: Could we scope this down slightly and just transform the config argument instead of the entire call expression? Then we leave `configPosition` as a concern only for `insertDebugNameIntoCallExpression`.

```typescript
function insertDebugNameIntoConfigObject(config: ts.Expression, debugName: string): ts.Expression {
  // ...
}
```

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+    // Do nothing if an identifier is used as the config object
+    // Ex -
+    // const defaultObject = { equals: () => false };
+    // signal(123, defaultObject)
```
Question: Do we want to potentially prepend `debugName` to this value like:

```typescript
const foo = signal(123, {debugName: 'foo', ...defaultObject});
```

AleksanderBodurri replied:
I think no, since this interferes with object equality.

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+
+    if (ts.isObjectLiteralExpression(callExpression.arguments[configPosition])) {
+      return insertDebugNameIntoExistingConfigObject(callExpression, debugName, configPosition);
+    }
```
Consider: Move the `ts.isIdentifier` and `ts.isObjectLiteralExpression` checks into `insertDebugNameIntoExistingConfigObject`. Even if those are `false`, there's still an existing config object (albeit, not an object literal) so that function is still valid. It can then choose whether or not to no-op.

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+  if (debugNameExists) {
+    return callExpression;
+  }
+  properties.unshift(
```
Suggestion: Include a comment to explain that we're putting `debugName` first in case a subsequent spread operator overwrites this value. That way we don't accidentally overwrite an explicitly provided `debugName` value which wasn't statically analyzable.

dgp1130 commented on packages/compiler-cli/test/ngtsc/debug_transform_spec.ts:
```diff
@@ -0,0 +1,1784 @@
+          const jsContents = env.getContents('test.js');
+          const builtContent = (await esbuild.transform(jsContents, minifiedProdBuildOptions)).code;
+          expect(builtContent).toContain(
+            `signal("Hello World", { equal: /* @__PURE__ */ __name(() => !0, "equal") });`,
```
Sugestion: Avoid depending on the specific output format of esbuild as much as possible (ex. `/* @__PURE__ */` or `!0`).

We could make this assertion a little more flexible with a regex or maybe adjust the input slightly like:

```typescript
import {signal} from '@angular/core';
declare function equal(): boolean;
const testSignal = signal('Hello World', { equal });
```

I suspect that would avoid any unnecessary transformations on `equal`.

dgp1130 commented on packages/compiler-cli/test/ngtsc/debug_transform_spec.ts:
Thanks for putting together all these comprehensive tests cases! I think this is super helpful to making sure this stays stable and covers all the edge cases.

josephperrott reviewed (approved):
LGTM

devversion commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/input_output_parse_options.ts:
```diff
@@ -11,16 +11,114 @@ import ts from 'typescript';
 import {ErrorCode, FatalDiagnosticError} from '../../../diagnostics';
 import {reflectObjectLiteral} from '../../../reflection';
 
+function checkInputForForDebugNameCase(
```
Why does the Angular compiler need to detect this pattern? Isn't the transform emitting this form _after_ the Angular compiler runs?

alxhub reviewed (commented):
First round of comments!

Additionally, in the commit history, use `refactor` commits for implementation details. A given feature (debug names for example) should have exactly 1 `feat` commit, which is the commit which actually makes the feature available for end users to use.

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
```diff
@@ -0,0 +1,450 @@
+  }
+
+  const specifier = importDeclaration.moduleSpecifier.text;
+  return specifier !== undefined && (specifier === '@angular/core' || specifier.includes('@angular/core/'));
```
Consider: The name `expressionIsUsingAngularImportedSymbol` implies that it's coming from any `@angular/*` package. We should either relax this check to `@angular/*` or update the function name to `...AngularCoreImportedSymbol`.

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
```diff
@@ -0,0 +1,450 @@
+ * - viewChildren
+ * - contentChild
+ * - contentChildren
+ * - effect
```
Consider: This list is very likely to fall out of date, I probably just dropping it to avoid misleading future readers. I don't think an exhaustive list is necessary. You can say "signal functions (`signal`, `computed`, `effect`, etc.)" and get the same point across.

dgp1130 commented on packages/compiler-cli/test/ngtsc/debug_transform_spec.ts:
```diff
@@ -0,0 +1,1784 @@
+
+const testFiles = loadStandardTestFiles();
+
+const DEBUG_NAME = 'debugName';
```
Consider: Since we've agreed on `debugName` and have landed other PRs with that name, is it still useful to have this constant? Should we just inline it everywhere?

dgp1130 commented on packages/compiler-cli/test/ngtsc/debug_transform_spec.ts:
```diff
@@ -0,0 +1,1784 @@
+                    import {signal, computed} from '@angular/core';
+                    const testSignal = signal(123);
+                    const testComputed = computed(() => testSignal(), { equal: () => true });
+                `,
```
Style: The indenting in these blocks seems kind of random/inconsistent. I suggest lining up the backticks at the same indentation level and putting the inline code one-level indented from there.

```typescript
{
  {
      {
        env.write(
          'test.ts',
          `
            import {signal, computed} from '@angular/core';
            // ...
          `
        );
      }
  }
}
```

alxhub commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
```diff
@@ -0,0 +1,448 @@
+      ]),
+      undefined,
+      ts.factory.createArrayLiteralExpression([
+        ts.factory.createObjectLiteralExpression(existingArgument.properties),
```
This is still a bit sketchy to me. This creates a node without any context (e.g. sourcemaps).

Ideally, we wouldn't have to copy the user's code like this, but we should still use `ts.update*` instead.

dgp1130 replied:
We discussed this briefly in the DevTools sync today. We can definitely update this object literal to just be `existingArgument`. How should this work for the duplicated node though?

We're effectively converting `signal(0, { equals })` to `signal(0, ...(ngDevMode ? [{ debugName: 'test', equals }] : [{ equals }]))`. That moves `{ equals }` to the `false` case, but also duplicate its in the `true` case. I presume we just need to `ts.factory.create*(...)` for the `true` case and can't update the existing one because we need it for the `false` case?

AleksanderBodurri replied:
Tested over the weekend with ng new and the compiler cli code with the transform. Source maps are working as expected even with the transformation applied. I think this is working as expected now

AleksanderBodurri commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
```diff
@@ -0,0 +1,450 @@
+  let expression = binaryExpression.right.expression;
+  if (ts.isPropertyAccessExpression(expression)) {
+    expression = expression.expression;
+  }
```
We decided to defer catching this case for now

AleksanderBodurri committed:
refactor(compiler-cli): implement transform to determine debugName from signal functions

Implements a compiler transform that attempts to statically analyze variable names and apply them to usages of signal functions like signal, computed, effect, etc.

dgp1130 commented:
I ran a [presubmit](http://fusion2/presubmit/715076116/OCL:715076116:BASE:715076143:1736801754914:95c9a31a/targets/invocations/33cc2698-7c57-494c-8a8f-47884c5ade0a/targets/%2F%2Fjava%2Fcom%2Fgoogle%2Ffirebase%2Fconsole:ui_build_test/log) and got some duplicate message errors. Debugging further, I see a problem with how we're generating this code.

Internally, i18n messages outside of Angular templates are defined via Closure's `goog.getMsg` API (roughly equivalent to Angular's `$localize`).

```typescript
/** @desc Say hello. */
const MSG_HELLO = goog.getMsg('Hello, Alex!');
console.log(MSG_HELLO); // 'Hello, Alex!' (or localized equivalent)
```

It's somewhat reasonable to parameterize such messages in a `computed` like so:

```typescript
const name = signal('Alex');
const hello = computed(() => {
  /** @desc Say hello. */
  const MSG_HELLO = goog.getMsg('Hello, {$name}', {name: name()});
  return MSG_HELLO;
});
```

Running the transform on this, the body of the computed gets duplicated like so:

```typescript
const name = signal(...(ngDevMode ? ['Alex', {debugName: 'name'}] : ['Alex']));
const hello = computed(...(ngDevMode ? [
  () => {
    /** @desc Say hello. */
    const MSG_HELLO = goog.getMsg('Hello, {$name}', {name: name()});
    return MSG_HELLO;
  },
  {debugName: 'hello'},
] : [
  () => {
    /** @desc Say hello. */
    const MSG_HELLO = goog.getMsg('Hello, {$name}', {name: name()});
    return MSG_HELLO;
  },
]);
```

Note that we now have two `goog.getMsg` calls. Closure Compiler extracts all i18n messages in the bundle by analyzing the AST, meaning it now extracts two `MSG_HELLO` symbols and errors because of the conflict. It would not be able to disambiguate between the two. I wonder if `$localize` might actually have a similar problem. :thinking:

We could ask Closure Compiler to be a little more lenient here, as the two messages are literally identical, it could probably make this work. However I do expect that erroring on duplicate messages is a generally reasonable thing for the compiler to do and I wouldn't want to turn that off just because of this case. Looking at this code it is also clear that we're duplicating a significant amount of code in the `computed` expression and dev builds typically run uncompiled, meaning both branches are shipped to the browser (even if only one is executed), so this is a potentially significant performance regression for developers, even if it ultimately gets compiled out for production users.

I remember we had some discussion about how safe it was to duplicate the expression in a signal and whether that might produce duplicated side effects. I think we ultimately decided it didn't because the ternary short circuits and only the true branch gets evaluated at runtime. However I think this issue makes it clear that duplicating the code at all, even if never executed, is still a potential issue.

I think the path forward here is to avoid duplicating the expression by putting it in a variable used in both branches of the `ngDevMode` ternary. My initial attempt was to use an IIFE using an inlined parameter like so:

```typescript
const name = signal(...(ngDevMode ? ['Alex', {debugName: 'name'}] : ['Alex']));
const hello = computed(...((msg) => ngDevMode ? [msg, {debugName: 'hello'}] : [msg])(() => {
  /** @desc Say hello. */
  const MSG_HELLO = goog.getMsg('Hello, {$name}', {name: name()});
  return MSG_HELLO;
});
```

This keeps a single definition of the function and only one `goog.getMsg` while ultimately evaluating to the same thing based on `ngDevMode`.

The problem with this is that it might be just a little too complicated an expression for bundlers. Testing this out I see:

```javascript
// Ideally:
computed(() => "Hello, Alex!");

// Closure Compiler:
computed(() => "Hello, Alex!");

// Angular CLI: esbuild + terser
computed(...(t=>[t])(()=>"Hello, Alex!"));
```

So Closure does seem smart enough to make this work, but esbuild and Terser are coming up a little short. I suspect the problem is that esbuild/Terser don't want to inline `t` in the IIFE because if `t` is used multiple times the expression would be duplicated. It's only safe to inline that variable because it's only used once in `t=>[t]`. Based on that, we can maybe rework this to explicitly put the computed function into a variable so the bundler doesn't need to inline an IIFE with a parameter. Second attempt:

```typescript
const hello = (() => {
  const x = () => 'Hello, Alex!';
  return computed(...(ngDevMode ? [x, {debugName: 'title'}] : [x]);
})();
```

This outputs:

```javascript
// Closure Compiler:
(() => computed(() => "Hello world"))();

// Angular CLI: esbuild + Terser
computed(()=>"Hello, World!");
```

So now we've got an extra IIFE in Closure which is failing to be inlined. It's a little better than this doesn't have a spread operator in the `computed` call, which is likely a runtime perf deoptimization as the engine probably needs to do extra work to expand the array.

I'm not sure why Closure Compiler can't inline that IIFE, especially since it inlined a much more complicated IIFE in my previous attempt. I'll ask around and see if we can find a pattern which optimizes ideally for both sets of bundlers.

dgp1130 commented:
Hmm, retrying the second solution I'm actually now seeing that it outputs `computed(() => 'Hello, Alex!')` in both cases, I'm not sure why my initial testing came up different.

If that's reliable then it seems like that transform would be valid path forward, I'll keep an eye out if I can replicate the extra IIFE I saw previously.

dgp1130 commented:
Global presubmit is fully passing :tada:, so I think this output format is looking good.

http://test/OCL:728286513:BASE:728422677:1739934312484:4bfbb98b

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
```diff
@@ -0,0 +1,473 @@
+  return false;
+}
+
+function transformVariableDeclaration(
```
Consider: Some short before/after examples in doc comments for the transform functions would be really helpful.

Doesn't need to be as involved as what you have in `signalMetadataTransform`, just a basic example would go a long way.

AleksanderBodurri replied:
These are included in `signalMetadataTransform`

dgp1130 commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
````diff
@@ -0,0 +1,473 @@
+ *
+ * becomes
+ * ```
+ * const mySignal = signal(...(ngDevMode ? ['Hello World', { debugName: "mySignal" }] : ['Hello World']));
````
Nit: I think we've now changed this to:

```suggestion
 * const mySignal = signal('Hello World', ...(ngDevMode ? [{ debugName: "mySignal" }] : []));
```

dgp1130 reviewed (commented):
Thanks for working through all these changes, it's looking great!

No major concerns on my side, just a few minor suggestions and questions. I think a couple existing comments are still open too when you get a chance.

AleksanderBodurri commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
```diff
@@ -0,0 +1,473 @@
+  // Do nothing if an identifier is used as the config object
+  // Ex -
+  // const defaultObject = { equals: () => false };
+  // signal(123, defaultObject)
```
Initially I was thinking this would be too magical because it would remove object equality, but for this specific case the config object itself is never referenced after the signal function is created. So this is probably fine, will investigate implementing this

AleksanderBodurri commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
```diff
@@ -0,0 +1,473 @@
+    return callExpression;
+  }
+
+  if (!ts.isObjectLiteralExpression(existingArgument)) {
```
Same as https://github.com/angular/angular/pull/57348#discussion_r1987338095

alxhub commented on packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts:
```diff
@@ -0,0 +1,473 @@
+  }
+
+  // climb up the tree from the import specifier to the import declaration
+  const importSpecifier = declarations[0];
```
No, multiple declarations would be something like:

```
interface Foo {
  bar: string;
}

interface Foo {
  baz: number;
}
```

dgp1130 replied:
Is it safe to just look at the first declaration? I guess if it is a real `signal` function then it will always have just one declaration from the `import`?

dgp1130 commented:
Passing TGP: http://test/OCL:749909832:BASE:750657379:1745431857014:f318b2d7

AleksanderBodurri commented on packages/compiler-cli/src/ngtsc/transform/src/debug_transform.ts:
```diff
@@ -0,0 +1,350 @@
+  if (newArgs.length > configPosition) {
+    newArgs[configPosition] = spreadElement;
+  } else {
+    newArgs.push(spreadElement);
```
I think this is a case we can account for when/if it arises. For now we have an allowlist of signal function that get transformed anyways so this is something we would have to explicitly add any new functions anyways

dgp1130 commented:
TGP still looking good: http://test/OCL:758440362:BASE:758744767:1747244383957:e4a093db

kirjs commented:
This PR was merged into the repository by commit 3a9a70de0829f69107ceaaa855d93281ecf791f1.

The changes were merged into the following branches: main
